### PR TITLE
Fixed: Multiple errors on single typo in match pattern

### PR DESCRIPTION
**Pull request by PramodBisht**
Here we have fixed the case where we were throwing two diagnostic messages `E0026` and `E0027` for same case.

Example
```
error[E0026]: variant `A::A` does not have a field named `fob`
  --> src/test/ui/issue-52717.rs:20:12
   |
20 |     A::A { fob } => { println!("{}", fob); }
   |            ^^^ variant `A::A` does not have this field

error[E0027]: pattern does not mention field `foo`
  --> src/test/ui/issue-52717.rs:20:5
   |
20 |     A::A { fob } => { println!("{}", fob); }
   |     ^^^^^^^^^^^^ missing field `foo`

error: aborting due to 2 previous errors
```

Here above we can see that both `E0026` and `E0027` are depicting
same thing.

So, to fix this issue, we are simply checking if for last element of `inexistent_fields` is there any value lies in `unmentioned_fields` using levenshtein algorithm, if it does then for that case we are simply deleting element from `unmentioned_fields`. More or less, now instead of showing separate message in `E0027` we are giving extra hint on `E0026`

r? @estebank

### CODE REVIEW

File: src/librustc_typeck/check/_match.rs [hunk 1]
```diff
@@ -944,13 +950,26 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
...
                                        kind_name,
                                        tcx.item_path_str(variant.did),
                                        t,
                                        plural));
+                if plural=="" {
+                    let input = unmentioned_fields.iter().map(|field|
+                                        Symbol::intern(field.to_string().as_str()))
+                                        .collect::<Vec<Symbol>>();
+                    let suggested_name = find_best_match_for_name(input.iter(),
+                                                                &ident.name.as_str(), None);
```
Comment (estebank) [inline src/librustc_typeck/check/_match.rs]: Fix indentation above.

**Review by estebank — APPROVED** [review]
Couple of nitpicks, but lgtm

File: src/librustc_typeck/check/_match.rs (hunk 1 above)
Comment (PramodBisht) [reply src/librustc_typeck/check/_match.rs]: which IDE or tools I should use to detect this issues?

Comment (estebank) [reply src/librustc_typeck/check/_match.rs]: Generally, you can make your changes, run `rustfmt` on the file and only take the changes for the code you changed :-/

That workflow is annoying enough that I just memorized the most relevant rules. Basically, if you have multiple lines

```
like(this,
     right,
     here);
```

you want them to be aligned to the `(`, which is why there's a preference to

```
this_other(
    indentation,
    of,
    multiple,
    args,
);
```

as it is independent of function name length.

File: src/librustc_typeck/check/_match.rs [hunk 2]
```diff
@@ -944,13 +950,26 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
...
                                            field_names);
-            if let Some((span, _)) = inexistent_fields.last() {
+            if let Some((span, ident)) = inexistent_fields.last() {
                 err.span_label(*span,
                                format!("{} `{}` does not have {} field{}",
                                        kind_name,
                                        tcx.item_path_str(variant.did),
                                        t,
                                        plural));
+                if plural=="" {
```
[the remark this answers is not in the document]
Comment (PramodBisht) [reply src/librustc_typeck/check/_match.rs]: done!

File: src/librustc_typeck/check/_match.rs [hunk 3]
```diff
@@ -944,13 +950,26 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
...
+                    let input = unmentioned_fields.iter().map(|field|
+                                        Symbol::intern(field.to_string().as_str()))
+                                        .collect::<Vec<Symbol>>();
```
[the remark this answers is not in the document]
Comment (PramodBisht) [reply src/librustc_typeck/check/_match.rs]: Fixed :)

[the remark this answers is not in the document]
Comment (PramodBisht) [reply src/librustc_typeck/check/_match.rs]: Thanks :)

File: src/librustc_typeck/check/_match.rs [hunk 4]
```diff
@@ -944,13 +950,26 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
...
+                    let suggested_name = find_best_match_for_name(input.iter(),
+                                                                &ident.name.as_str(), None);
+                    if let Some(suggested_name) = suggested_name {
+                        err.span_suggestion(*span, "did you mean", suggested_name.to_string());
+                        // we don't want to throw `E0027` in case we have thrown `E0026` for them
+                        unmentioned_fields.retain(|&x| x.to_string() !=
+                                                            suggested_name.to_string());
```
[the remark this answers is not in the document]
Comment (PramodBisht) [reply src/librustc_typeck/check/_match.rs]: done!

File: src/test/ui/issue-52717.rs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+enum A {
+  A {
+    foo: usize,
+  }
+}
+
+fn main() {
+  let x = A::A { foo: 3 };
+  match x {
+    A::A { fob } => { println!("{}", fob); }
```
Comment (estebank) [inline src/test/ui/issue-52717.rs]: There's another case that we're not handling (but it's fine):

```
enum A {
  A {
    foo: usize,
  }
}
 fn main() {
  let x = A::A { foo: 3 };
  match x {
    A::A { fob } => { println!("{}", foo); }
  }
}
```

After applying the suggestion, the code will complain about the printed `fob`.

It'll be interesting coming up with a solution for this as a follow up.

Comment (PramodBisht) [reply src/test/ui/issue-52717.rs]: right!

Comment (PramodBisht) [reply src/test/ui/issue-52717.rs]: Working toward that :)

**Pull request merged**